Quartz/Engine/Include/Quartz/Utilities/Commander.hpp

```diff
@@ -0,0 +1,116 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <string>
+#include <array>
+#include <functional>
+
+namespace qz
+{
+	namespace utils
+	{
+        constexpr int MaxCommandsNumber = 100;
+        constexpr int MaxArgumentNumber = 10;
+
+        /**
+         * @brief The command book stores commands and information on them to be used by a commander
+         */
+
+        struct CommandBook{
+            std::array<std::string, MaxCommandsNumber> m_command;
+            std::array<std::string, MaxCommandsNumber> m_help;
+            std::array<std::string, MaxCommandsNumber> m_permission;
+            std::array<std::function<int(std::array<std::string, MaxArgumentNumber> args)>, MaxCommandsNumber> m_functions;
+
+            /**
+             * @brief Registers a command in the command registry
+             * 
+             * @param command The keyword for calling the command
+             * @param permission What permission is required to run this command
+             */
+            int reg(const std::string& command, const std::string& help, const std::string& permission, std::function<int(std::array<std::string, MaxArgumentNumber> args)> f);
+    
+            /**
+             * @brief Searches for a command and returns the index its stored at
+             */
+            int find(const std::string& command);
+
+            /**
+             * @brief Gets next open space in book
+             */
+            int i();
+
+        private:
+            int m_i;
+
+        };
+
+        /**
+         * @brief The Commander handles comand line functions while the server is running. A loop will watch for commands from the terminal while the server runs but the clients can also send commands.
+         * 
+         */
+        class Commander{
+        private:
+            CommandBook m_book;
+            std::ostream& m_out;
+            std::istream& m_in;
+
+        public:
+            Commander(CommandBook& book, std::ostream& out, std::istream& in);
+            ~Commander();
+
+            /**
+             * @brief Calls a command 
+             * 
+             * @param command The keyword for calling the command.
+             */
+            int run(const std::string& command, const std::array<std::string, MaxArgumentNumber> args);
+
+            /**
+             * @brief Returns helpstring for command
+             * 
+             * @param args array of input, args[0] is the command helpstring is returned for
+             */
+            int help(std::array<std::string, qz::utils::MaxArgumentNumber> args);
+
+            /**
+             * @brief Returns string listing available commands 
+             */
+            int list();
+
+            /**
+             * @brief Listens for commands.
+             */
+            int post();
+
+            int initialize();
```

Comment:
If you could document in the doxygen comments what these return values are? (expected values on success, expected values on error, generally what they represent).